src/SourceBrowser.Site/Controllers/UploadController.cs

```diff
@@ -31,64 +31,86 @@ public ActionResult Submit(string githubUrl)
-                htmlTransformer.Visit(workspaceModel);
+                var organizationPath = System.Web.Hosting.HostingEnvironment.MapPath("~/") + "SB_Files\\" + retriever.UserName;
+                var repoPath = Path.Combine(organizationPath, retriever.RepoName);
 
-                var searchTransformer = new SearchIndexTransformer(retriever.UserName, retriever.RepoName);
-                searchTransformer.Visit(workspaceModel);
+                // TODO: Use parallel for.
+                // TODO: Process all solutions.
+                // For now, we're assuming the shallowest and shortest .sln file is the one we're interested in
+                foreach (var solutionPath in solutionPaths.OrderBy(n => n.Length).Take(1))
+                {
+                    Generator.Model.WorkspaceModel workspaceModel;
+                    try
+                    {
+                        workspaceModel = UploadRepository.ProcessSolution(solutionPath, repoRootPath);
+                    }
+                    catch (Exception ex)
+                    {
+                        ViewBag.Error = "There was an error processing solution " + Path.GetFileName(solutionPath);
+                        return View("Index");
+                    }
 
+                    //One pass to lookup all declarations
+                    var typeTransformer = new TokenLookupTransformer();
+                    typeTransformer.Visit(workspaceModel);
+                    var tokenLookup = typeTransformer.TokenLookup;
 
+                    //Another pass to generate HTMLs
+                    var htmlTransformer = new HtmlTransformer(tokenLookup, repoPath);
+                    htmlTransformer.Visit(workspaceModel);
 
-                // Generate HTML of the tree view
-                var treeViewTransformer = new TreeViewTransformer(repoPath, retriever.UserName, retriever.RepoName);
-                treeViewTransformer.Visit(workspaceModel);
-            }
+                    var searchTransformer = new SearchIndexTransformer(retriever.UserName, retriever.RepoName);
+                    searchTransformer.Visit(workspaceModel);
 
-            return Redirect("/Browse/" + retriever.UserName + "/" + retriever.RepoName);
+                    // Generate HTML of the tree view
+                    var treeViewTransformer = new TreeViewTransformer(repoPath, retriever.UserName, retriever.RepoName);
+                    treeViewTransformer.Visit(workspaceModel);
+                }
+
+                processingSuccessful = true;
+                return Redirect("/Browse/" + retriever.UserName + "/" + retriever.RepoName);
+            }
+            finally
+            {
+                if (processingSuccessful)
+                {
+                    BrowserRepository.MarkRepositoryReady(retriever.UserName, retriever.RepoName);
+                }
+                else
+                {
+                    BrowserRepository.UnlockRepository(retriever.UserName, retriever.RepoName);
```

Comment:
Great point. That'd simplify the `BrowseController` and I'd love it to be this way, but we have an issue with the libgit2 library:
If something goes wrong, we should remove the `SB_Files/UserName/RepoName' directory to designate it as non-existent. However, libgit2 holds a lock on that directory (particularly on the `.git` hidden subdirectory.)
If we could figure out how to tackle this, then that'd be great.